src/encoding/schema/map.ts

```diff
@@ -0,0 +1,144 @@
+import { Schema, MsgpackEncodingData, JSONEncodingData } from '../encoding.js';
+
+/* eslint-disable class-methods-use-this */
+
+export interface NamedMapEntry {
+  key: string;
+  valueSchema: Schema;
+  required: boolean;
+  omitEmpty: boolean;
+}
+
+export function allOmitEmpty(
+  entries: Array<Omit<NamedMapEntry, 'omitEmpty'>>
+): NamedMapEntry[] {
+  return entries.map((entry) => ({ ...entry, omitEmpty: true }));
+}
+
+export class NamedMapSchema extends Schema {
+  constructor(public readonly entries: NamedMapEntry[]) {
+    super();
+  }
+
+  public defaultValue(): Map<string, unknown> {
+    const map = new Map<string, unknown>();
+    for (const entry of this.entries) {
+      if (entry.required) {
+        map.set(entry.key, entry.valueSchema.defaultValue());
+      }
+    }
+    return map;
+  }
+
+  public isDefaultValue(data: unknown): boolean {
+    if (!(data instanceof Map)) return false;
+    for (const entry of this.entries) {
+      if (data.has(entry.key) && data.get(entry.key) !== undefined) {
+        if (
+          !entry.required ||
+          !entry.valueSchema.isDefaultValue(data.get(entry.key))
+        ) {
+          return false;
+        }
+      }
+    }
+    // TODO: examine if there are any extra keys?
```

Comment:
how urgent is this?

Reply:
I looked into it, and ignoring extra keys seems to make the most sense. Comment removed and unit tests with extra keys added in f79528e68ce60e0a1142fab4c67e6ebdb4379caf